sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypes.scala

```diff
@@ -62,9 +62,14 @@ case class CreateStruct(children: Seq[NamedExpression]) extends Expression {
   override lazy val dataType: StructType = {
     assert(resolved,
       s"CreateStruct contains unresolvable children: ${children.filterNot(_.resolved)}.")
-    val fields = children.map { child =>
-      StructField(child.name, child.dataType, child.nullable, child.metadata)
-    }
+      val fields = children.zipWithIndex.map { case (child, idx) =>
+        child match {
+          case ne: NamedExpression =>
```

Comment:
I will not argue that they have some share logic, however, the difference between `GenericUDFStruct ` and `GenericUDFNamedStruct` is, the former gives names for output fields as (col1, col2 ...), but the later requires specifying the output fields names from its arguments. 
Probably we'd better create another expression class in Spark SQL for `GenericUDFNamedStruct`, mixing the functionality in a single class probably makes people confusing.

It will be great if you can create an issue under the expression umbrella task https://issues.apache.org/jira/browse/SPARK-8159 

What do you think?